File: csharp/DataStructures/src/dsa/sorting/SelectionSort.cs
namespace dsa.sorting;

public class SelectionSort
{
    // static void Main()
    // {
    //     int[] arr = new int[] { 5, 1, 2, 9, 10 };

    //     SelectionSort ss = new();

    //     ss.printArray(arr);
    //     ss.sort(arr);
    //     ss.printArray(arr);
    // }

    public void printArray(int[] arr)
    {
        int n = arr.Length;

        for (int i = 0; i < n; i++)
        {
            Console.Write(arr[i] + " ");
        }
        Console.WriteLine();
    }

    public void sort(int[] arr)
    {
        int n = arr.Length;

        // Percorre cada elemento do array
        for (int i = 0; i < n - 1; i++)
        {
            // Assume que o elemento atual é o menor
            int min = i;

            // Procura pelo menor elemento no restante do array
            for (int j = i + 1; j < n; j++)
            {
                if (arr[j] < arr[min])
                    min = j;
            }

            // Troca o elemento atual com o menor elemento encontrado
            int temp = arr[min];
            arr[min] = arr[i];
            arr[i] = temp;
        }
    }
}

File: go/dsa/sorting/SelectionSort/SelectionSort.go
package selectionsort

import "fmt"

// PrintArray imprime os elementos do array em uma única linha.
func PrintArray(arr []int) {
	for _, v := range arr { // Itera sobre cada elemento do array
		fmt.Printf("%d ", v) // Imprime o valor atual com um espaço
	}
	fmt.Println() // Quebra de linha após imprimir todos os elementos
}

// Sort realiza o algoritmo de ordenação Selection Sort em um array.
func Sort(arr []int) {
	n := len(arr) // Obtém o tamanho do array

	// Laço externo que percorre o array até o penúltimo elemento
	for i := 0; i < n-1; i++ {

		min := i // Assume que o índice atual é o menor

		// Laço interno que percorre os elementos restantes
		for j := i + 1; j < n; j++ {

			// Se encontrar um valor menor que o valor atual mínimo,
			// atualiza o índice do menor valor.
			if arr[j] < arr[min] {
				min = j
			}
		}

		// Troca o menor valor encontrado com o valor da posição inicial do laço externo
		temp := arr[min]  // Armazena o menor valor temporariamente
		arr[min] = arr[i] // Move o valor da posição inicial para a posição do menor valor
		arr[i] = temp     // Coloca o menor valor na posição inicial
	}
}
